airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/AbstractJdbcDestination.java

```diff
@@ -0,0 +1,185 @@
+import io.airbyte.integrations.base.TruncateInsertIntoConsumer;
+import io.airbyte.protocol.models.AirbyteConnectionStatus;
+import io.airbyte.protocol.models.AirbyteConnectionStatus.Status;
+import io.airbyte.protocol.models.AirbyteMessage;
+import io.airbyte.protocol.models.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.ConnectorSpecification;
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.Map;
+import org.jooq.DSLContext;
+import org.jooq.SQLDialect;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractJdbcDestination implements Destination, TableCreationOperations, InsertTableOperations, BufferedWriteOperations {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AbstractJdbcDestination.class);
+  static final String COLUMN_NAME = "data";
+
+  private final String driverClass;
+  private final SQLDialect dialect;
+  private final SQLNamingResolvable namingResolver;
+  private Database database;
+
+  public AbstractJdbcDestination(final String driverClass, final SQLDialect dialect, final SQLNamingResolvable namingResolver) {
+    this.driverClass = driverClass;
+    this.dialect = dialect;
+    this.namingResolver = namingResolver;
+  }
+
+  @Override
+  public ConnectorSpecification spec() throws IOException {
+    // return a JsonSchema representation of the spec for the integration.
+    final String resourceString = MoreResources.readResource("spec.json");
+    return Jsons.deserialize(resourceString, ConnectorSpecification.class);
+  }
+
+  @Override
+  public AirbyteConnectionStatus check(JsonNode config) {
+    try (final Database database = getDatabase(config)) {
+      // attempt to get current schema. this is a cheap query to sanity check that we can connect to the
+      // database. `currentSchema()` is a jooq method that will run the appropriate query based on which
+      // database it is connected to.
+      database.query(this::getCurrentDatabaseName);
+
+      return new AirbyteConnectionStatus().withStatus(Status.SUCCEEDED);
+    } catch (Exception e) {
+      LOGGER.debug("Exception while checking connection: ", e);
+      return new AirbyteConnectionStatus()
+          .withStatus(Status.FAILED)
+          .withMessage("Can't connect with provided configuration.");
+    }
+  }
+
+  protected Database getDatabase(JsonNode config) {
+    final JsonNode jdbcConfig = toJdbcConfig(config);
+
+    return Databases.createDatabase(
+        jdbcConfig.get("username").asText(),
+        jdbcConfig.has("password") ? jdbcConfig.get("password").asText() : null,
+        jdbcConfig.get("jdbc_url").asText(),
+        driverClass,
+        dialect);
+  }
+
+  public abstract JsonNode toJdbcConfig(JsonNode config);
+
+  protected String getCurrentDatabaseName(DSLContext ctx) {
+    return ctx.select(currentSchema()).fetch().get(0).getValue(0, String.class);
+  }
+
+  @Override
+  public SQLNamingResolvable getNamingResolver() {
```

Comment:
why is this public? can it be private? i'm only seeing it used on line 127 and then in tests. okay to do `@VisibleForTesting` if it's necessary, but since it is dependency injected, i think it should be needed?

Reply:
it's public because it comes from the interface, if i put private it doesn't compile?

```
'getNamingResolver()' in 'io.airbyte.integrations.destination.jdbc.AbstractJdbcDestination' clashes with 'getNamingResolver()' in 'io.airbyte.integrations.base.Destination'; attempting to assign weaker access privileges ('private'); was 'public'
```

Reply:
okay. i guess i'm a bit confused about why it's in the destination interface at all. it seems like we only use it internally to destination and don't really intend to use it externally (at least for now). we can figure this out later.